src/Temporalio/Worker/ActivityWorker.cs

```diff
@@ -0,0 +1,703 @@
+        /// Finalizes an instance of the <see cref="ActivityWorker"/> class.
+        /// </summary>
+        ~ActivityWorker()
+        {
+            Dispose(false);
+        }
+
+        /// <summary>
+        /// Execute the activity until poller shutdown or failure.
+        /// </summary>
+        /// <returns>Task that only completes successfully on poller shutdown.</returns>
+        public async Task ExecuteAsync()
+        {
+            // Run poll loop until there is no poll left
+            using (logger.BeginScope(new Dictionary<string, object>()
+            {
+                ["TaskQueue"] = worker.Options.TaskQueue!,
+            }))
+            {
+                while (true)
+                {
+                    var task = await worker.BridgeWorker.PollActivityTaskAsync().ConfigureAwait(false);
+                    logger.LogTrace("Received activity task: {Task}", task);
+                    switch (task?.VariantCase)
+                    {
+                        case Bridge.Api.ActivityTask.ActivityTask.VariantOneofCase.Start:
+                            StartActivity(task);
+                            break;
+                        case Bridge.Api.ActivityTask.ActivityTask.VariantOneofCase.Cancel:
+                            if (!runningActivities.TryGetValue(task.TaskToken, out var act))
+                            {
+                                logger.LogWarning(
+                                    "Cannot find activity to cancel for token {TaskToken}",
+                                    task.TaskToken);
+                            }
+                            else
+                            {
+                                act.Cancel(task.Cancel.Reason);
+                            }
+                            break;
+                        case null:
+                            // This means worker shut down
+                            return;
+                        default:
+                            throw new InvalidOperationException(
+                                $"Unexpected activity task case {task?.VariantCase}");
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// Gracefully shutdown this worker after <see cref="ExecuteAsync" /> is complete. This must
+        /// not be called until <see cref="ExecuteAsync" /> has returned. This should never throw.
+        /// </summary>
+        /// <returns>
+        /// Task that completes when all activities have completed. If a running activity does not
+        /// respond to cancellation, this may never return.
+        /// </returns>
+        public async Task GracefulShutdownAsync()
+        {
+            using (logger.BeginScope(new Dictionary<string, object>()
+            {
+                ["TaskQueue"] = worker.Options.TaskQueue!,
+            }))
+            {
+                workerShutdownTokenSource.Cancel();
+                logger.LogInformation(
```

Comment:
Maybe log before cancel?

Reply:
Cancel is a synchronous notification so doesn't _really_ matter, but will update

Reply:
Fixed